### ISSUE TITLE
feat: rewrite Mutex/Semaphore

### ISSUE THREAD

**Issue by PandaWorker** [jsr-core/asyncutil#30]
```ts
class DequeItem<T> {
	value: T;
	prev: DequeItem<T> | null = null;
	next: DequeItem<T> | null = null;

	constructor(value: T) {
		this.value = value;
	}
}

export class Deque<T = unknown> {
	#size: number = 0;
	#head: DequeItem<T> | null = null;
	#tail: DequeItem<T> | null = null;

	get size() {
		return this.#size;
	}

	get head() {
		return this.#head;
	}

	get tail() {
		return this.#tail;
	}

	get isEmpty(): boolean {
		return this.size === 0;
	}

	static from<U>(collection: ArrayLike<U> | Iterable<U>) {
		const arr = Array.from(collection);
		const list = new Deque<U>();

		for (const elem of arr) {
			list.push(elem);
		}

		return list;
	}

	push(...values: T[]) {
		for (const value of values) {
			const newItem = new DequeItem(value);

			if (this.isEmpty) {
				this.#head = this.#tail = newItem;
			} else {
				newItem.prev = this.#tail;
				this.#tail!.next = newItem;
				this.#tail = newItem;
			}
			this.#size++;
		}
	}

	unshift(...values: T[]) {
		for (const value of values) {
			const newItem = new DequeItem(value);

			if (this.isEmpty) {
				this.#head = this.#tail = newItem;
			} else {
				newItem.next = this.#head;
				this.#head!.prev = newItem;
				this.#head = newItem;
			}
			this.#size++;
		}
	}

	pop(): T | null {
		if (this.isEmpty) return null;

		const value = this.#tail!.value;
		this.#tail = this.#tail!.prev;

		if (this.#tail) {
			this.#tail.next = null;
		} else {
			this.#head = null;
		}

		this.#size--;
		return value;
	}

	shift(): T | null {
		if (this.isEmpty) return null;

		const value = this.#head!.value;
		this.#head = this.#head!.next;

		if (this.#head) {
			this.#head.prev = null;
		} else {
			this.#tail = null;
		}

		this.#size--;
		return value;
	}

	clear() {
		this.#size = 0;
		this.#head = null;
		this.#tail = null;
	}

	[Symbol.iterator](): Iterator<T> {
		var current = this.#head;

		return {
			next: () => {
				if (current) {
					const value = current.value;
					current = current.next;
					return { value, done: false };
				} else {
					return { value: null, done: true };
				}
			},
		};
	}
}

export class Semaphore {
	#value: number;
	#waiters: Deque<PromiseWithResolvers<void>> = new Deque();

	get locked() {
		return this.#value !== 0;
	}

	constructor(value: number) {
		this.#value = value;
	}

	acquire(): Promise<Disposable> {
		const disposable = {
			[Symbol.dispose]: () => this.release(),
		};

		if (this.#value > 0) {
			this.#value -= 1;
			return Promise.resolve(disposable);
		}

		const waiter = Promise.withResolvers<void>();
		this.#waiters.push(waiter);

		return waiter.promise.then(() => disposable);
	}

	async runExclusive<R>(fn: () => PromiseLike<R> | R): Promise<R> {
		const lock = await this.acquire();
		return Promise.resolve(fn()).finally(() => lock[Symbol.dispose]());
	}

	release() {
		if (this.#waiters.size === 0) {
			this.#value += 1;
		} else {
			this.#waiters.shift()!.resolve();
		}
	}
}

export class Mutex extends Semaphore {
	constructor(){
		super(1);
	}
}

export class Lock <T> {
	#mu = new Mutex();
	#value: T;

	constructor(value: T){
		this.#value = value;
	}

	async lock<R>(fn: (value: T) => R | PromiseLike<R>): Promise<R> {
		const lock = await this.#mu.acquire();
		return Promise.resolve(fn(this.#value)).finally(() => lock[Symbol.dispose]());
	}

}
```

**Comment by lambdalisue** [comment]
I'm sorry but I don't get the reason.

**Comment by PandaWorker** [comment]
<img width="802" alt="image" src="https://github.com/user-attachments/assets/0b951e28-78f2-4109-844d-b5bd64465fe8">


code: https://gist.github.com/PandaWorker/ec07a22b55668073e70237285da18cee

**Comment by PandaWorker** [comment]
if you run the "task" function with all mutexes, it will work only for MyMutex
This is most likely due to competitive liberation.

```ts
const subtask = async (mutex: {acquire: () => Promise<Disposable>}) => {
	const lock = await mutex.acquire();
	lock[Symbol.dispose]();
}

const task = async (mutex: {acquire: () => Promise<Disposable>}, iterations: number = 10)  => {
	const tasks = [];
	for (let i = 0; i < iterations; i++) {
		tasks.push(subtask(mutex));
	}
	await Promise.all(tasks);
}
```

**Comment by mdekstrand** [comment]
I'm not sure that introducing a whole new data structure is terribly helpful, but I do think there's some significant opportunity to speed up `Mutex`, and basing it on `Semaphore` to share logic makes sense to me — a mutex is a semaphore with count 1.

I was separately looking through the `Mutex` code today for some performance review of one of my projects, and I'm not sure why the current implementation needs to wait on the entire list of previous waiters each time it is acquired. This code doesn't do that, which should be part of the speedup. Not allocating a lot of new arrays is likely another part.

If it's acceptable for the mutex to wake up the most recent locker first, instead of the oldest locker first, then similar speedups can likely be achieved without the `Deque` by just pushing and popping from the end of an array. Mutexes don't typically make any guarantees about the order in which waiters awaken, so it shouldn't be a problem from a theoretical perspective.

**Comment by PandaWorker** [comment]
Here is a more or less optimized Mutex, but still Mutex using the Deque data structure is 10-20% faster

```ts
export class Semaphore {
	#waiters = new Set<PromiseWithResolvers<void>>();
	#value: number;

	get locked() {
		return this.#value < 0;
	}

	get waiters() {
		return this.#waiters.size;
	}

	constructor(value: number) {
	        if (value <= 0 || !Number.isSafeInteger(value)) {
			throw new RangeError(
				`value must be a positive safe integer, got ${value}`
			);
		}

		this.#value = value;
	}

	acquire() {
		const disposable = {
			[Symbol.dispose]: () => this.release(),
		};

		if (this.#value-- > 0) {
			return Promise.resolve().then(() => disposable);
		}

		const waiter = Promise.withResolvers<void>();
		this.#waiters.add(waiter);

		return waiter.promise.then(() => disposable);
	}

	release() {
		const waiters = this.#waiters;

		if (waiters.size === 0) {
			this.#value += 1;
		} else {
			// TODO:
			const iter = waiters.entries().take(1);
			const waiter = iter.next().value![0];
			waiter.resolve();
		}
	}
}

export class Mutex extends Semaphore {
	constructor() {
		super(1);
	}
}
```

<img width="803" alt="image" src="https://github.com/user-attachments/assets/6dc627b2-6bbb-4286-885d-2c2b95f9f242">

**Comment by PandaWorker** [comment]
That's what we managed to squeeze the maximum performance out of using Set

```ts
export class Semaphore {
	#waiters = new Set<PromiseWithResolvers<void>>();
	#value: number;

	get locked() {
		return this.#value === 0;
	}

	get waiters() {
		return this.#waiters.size;
	}

	constructor(value: number) {
		this.#value = value;
	}

	acquire() {
		const disposable = {
			[Symbol.dispose]: () => this.release(),
		};

		if (this.#value > 0) {
			this.#value -= 1;
			return Promise.resolve().then(() => disposable);
		}

		const waiter = Promise.withResolvers<void>();
		this.#waiters.add(waiter);

		return waiter.promise.then(() => disposable);
	}

	release() {
		if (this.#waiters.size > 0) {
			const waiters = this.#waiters;
			const [waiter] = waiters.keys();

			waiters.delete(waiter);
			waiter.resolve();
		} else {
			this.#value += 1;
		}
	}
}

export class Mutex extends Semaphore {
	constructor() {
		super(1);
	}
}
```

**Comment by lambdalisue** [comment]
So basically performance improvement, right? I'll try to implement.

**Comment by lambdalisue** [comment]
> if you run the "task" function with all mutexes, it will work only for MyMutex
This is most likely due to competitive liberation.

What did you mean by "work"? I think the code "work" well with current implementation.

```ts
import { Mutex } from "./mutex.ts";
const subtask = async (mu: Mutex, i: number) => {
  using _lock = await mu.acquire();
  console.log(`subtask ${i}`);
};
const tasks = [];
const mu = new Mutex();
for (let i = 0; i < 10; i++) {
  tasks.push(subtask(mu, i));
}
await Promise.all(tasks);
```

```
subtask 0
subtask 1
subtask 2
subtask 3
subtask 4
subtask 5
subtask 6
subtask 7
subtask 8
subtask 9
```

**Comment by lambdalisue** [comment]
@mdekstrand @PandaWorker Could you check https://github.com/jsr-core/asyncutil/pull/31

**Comment by lambdalisue** [comment]
Closed via #34
